src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <vector>
+#include <variant>
+
+namespace raksha::ir::auth_logic {
+
+class AstPrincipal {
+  public:
+    AstPrincipal(std::string name) : name_(std::move(name)) {}
+  private:
+    std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum AstSign { negated, positive };
+
+class AstPredicate {
+  public:
+    AstPredicate(
+        std::string name,
+        std::vector<std::string> args,
+        AstSign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        sign_(std::move(sign)) {}
+  private:
+    std::string name_;
+    std::vector<std::string> args_;
+    AstSign sign_;
+};
+
+// This represents an expression that applies either a predicate or a 
+// "canActAs" to a principal, and it is a sub-term of AstFlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class AstVerbPhrase {
+  public:
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPredicate pred);
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPrincipal right_prin);
+  private:
+    // The principal variant of the internal form means this represents
+    // "canActAs principal".
+    using InternalFormType = std::variant<AstPredicate, AstPrincipal>;
+    InternalFormType internal_form_;
+    AstVerbPhrase(InternalFormType internal_form)
+      : internal_form_(std::move(internal_form)) {}
+};
+
+// This the AST Node for FlatFact which takes on one of two forms:
+//   - <principal> <verb_phrase>  (* a principal applied to a verbphrase *)
+//   - <predicate>                (* just a predicate *)
+// A similar variant-based pattern is used for this as for VerbPhrase. Here an 
+// inner struct is used for the first form (principal applied to verb_phrase).
+class AstFlatFact {
+  public:
+    std::unique_ptr<AstFlatFact>
+      CreateFlatFact(AstPrincipal prin, AstVerbPhrase verb_phrase);
+    std::unique_ptr<AstFlatFact>
+      CreateFlatFact(AstPredicate prin);
+  private:
+    typedef struct VerbPhraseFlatFact {
+      AstPrincipal prin;
+      AstVerbPhrase verb_phrase;
+      VerbPhraseFlatFact(AstPrincipal prin, AstVerbPhrase verb_phrase)
+        : prin(std::move(prin)),
+          verb_phrase(std::move(verb_phrase)) {}
+    } VerbPhraseFlatFact;
```

Comment:
This pattern of `typedef struct <definition> StructName` is a very C way of defining a struct. In C, type aliases and struct names do not occupy the same name universe and you must qualify every `struct` name with `struct`. The pattern here introduces the `struct` name in both universes so that the `struct` qualifier is unnecessary.

C++ does not have this issue: you can describe `struct` and `class` type names without the additional `struct` qualifier. Therefore, this is unnecessary and non-idiomatic. In addition, it violates the rule about i[nternal aliases in headers](https://google.github.io/styleguide/cppguide.html#Aliases).

Reply:
fixed.